### PR TITLE
New apacheauth module

### PR DESCRIPTION
**Pull request by vccu**
Authenticates users against an Apache auth file.

### CODE REVIEW

File: 2.0/m_apacheauth.cpp [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+	}
+
+	bool SplitRecord(const std::string &record, std::string &record_algorithm, std::string &record_salt, std::string &record_hash)
+	{
+		const char *s = record.c_str();
+		const char *ptr;
+
+		if (record[0]!='$') return false;
+		s = s+1;
+		ptr = strchr(s, '$');
```
Comment (SaberUK) [inline 2.0/m_apacheauth.cpp]: You can use std::string::find here instead of using `const char*` and `strchr` here.

File: 2.0/m_apacheauth.cpp [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+		return true;
+	}
+
+	void LoadAuthFile()
+	{
+		logins.clear();
+		char s[2048];
+		int line=0;
+		ServerInstance->SNO->WriteGlobalSno('a', "Loading auth file %s",authfile.c_str());
+		FILE *f = fopen(authfile.c_str(), "r");
```
Comment (SaberUK) [inline 2.0/m_apacheauth.cpp]: Using std::ifstream might make this code cleaner.

File: 2.0/m_apacheauth.cpp [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+		std::map<std::string,std::string>::const_iterator i = logins.find(user->ident);
+		if (i == logins.end()) {
+			ServerInstance->SNO->WriteGlobalSno('a', "Forbiding connection from %s!%s@%s (login not found)",
+				user->nick.c_str(), user->ident.c_str(), user->host.c_str());
+			ServerInstance->Users->QuitUser(user, killreason);
+			return MOD_RES_PASSTHRU;
+		}
+
+		std::string record_algorithm, record_salt, record_hash;
+		if (!SplitRecord(i->second, record_algorithm, record_salt, record_hash)) {
```
Comment (SaberUK) [inline 2.0/m_apacheauth.cpp]: You should probably do this once on load and keep it in a custom structure.

File: 2.0/m_apacheauth.cpp [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+			return MOD_RES_PASSTHRU;
+		}
+		ServerInstance->SNO->WriteGlobalSno('a', "Granting access to connection from %s!%s@%s",
+				user->nick.c_str(), user->ident.c_str(), user->host.c_str());
+		return MOD_RES_PASSTHRU;
+	}
+
+	Version GetVersion()
+	{
+		return Version("Allow/Deny connections based upon an Apache auth file", VF_VENDOR);
```
Comment (SaberUK) [inline 2.0/m_apacheauth.cpp]: This is not a vendor module. Don't use VF_VENDOR.

**Comment by vccu** [comment]
All done. I pushed -f to the same branch, not sure this is what I'm meant to do, or push a new branch instead. Let me know if I should do the second way.

File: 2.0/m_apacheauth.cpp [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+		std::ifstream f;
+
+		logins.clear();
+		ServerInstance->SNO->WriteGlobalSno('a', "Loading auth file %s",authfile.c_str());
+		f.open(authfile.c_str(), std::ifstream::in);
+		if (!f.good()) {
+			ServerInstance->SNO->WriteGlobalSno('a', "Auth file failed to open, no connections will be allowed");
+			return;
+		}
+		while (1) {
```
Comment (SaberUK) [inline 2.0/m_apacheauth.cpp]: You can do something like this and avoid using a C buffer and an infinite loop.

```c++
std::string line;
while (std::getline(stream, line))
{
    // do something with line here.
}

**Comment by vccu** [comment]
Done, pushed -f again.

File: 2.0/m_apacheauth.cpp [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+		logins.clear();
+		ServerInstance->SNO->WriteGlobalSno('a', "Loading auth file %s",authfile.c_str());
+		f.open(authfile.c_str(), std::ifstream::in);
+		if (!f.good()) {
+			ServerInstance->SNO->WriteGlobalSno('a', "Auth file failed to open, no connections will be allowed");
+			return;
+		}
+		std::string s;
+		while (std::getline(f, s)) {
+			if (!f.good()) break;
```
Comment (SaberUK) [inline 2.0/m_apacheauth.cpp]: I think this is unnecessary now.

File: 2.0/m_apacheauth.cpp [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+			++line;
+			if (!s[0] || s[0] == '\r' || s[0] == '\n') continue;
```
Comment (SaberUK) [inline 2.0/m_apacheauth.cpp]: Same for this.

**Comment by vccu** [comment]
Changed and pushed -f again.

File: 2.0/m_apacheauth.cpp [hunk 8]
```diff
@@ -0,0 +1,159 @@
+/*       +------------------------------------+
```
Comment (rburchell) [inline 2.0/m_apacheauth.cpp]: this header style changed a while ago. it'd be nice if you could use the newer style (and your own copyright info)

File: 2.0/m_apacheauth.cpp [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+		if (!tag->getBool("useapacheauth", true))
+			return MOD_RES_PASSTHRU;
+
+		if (!allowpattern.empty() && InspIRCd::Match(user->ident,allowpattern))
+			return MOD_RES_PASSTHRU;
+
+		//ServerInstance->SNO->WriteGlobalSno('a', "Got request to auth %s with hash %s", user->nick.c_str(), user->password.c_str());
+		std::map<std::string, HashedPassword>::const_iterator i = logins.find(user->ident);
+		if (i == logins.end()) {
+			ServerInstance->SNO->WriteGlobalSno('a', "Forbiding connection from %s!%s@%s (login not found)",
```
Comment (rburchell) [inline 2.0/m_apacheauth.cpp]: typo in message, I'd suggest "Denying connection" instead perhaps.

while you're here, can I suggest that you remove the commented out lines?

**Comment by vccu** [comment]
Done.

**Pull request merged**